env/env_encrypt2.cc

```diff
@@ -0,0 +1,989 @@
+      uint64_t block_index = file_offset / BlockSize();
+
+      // make a context once per thread
+      if (!aes_context) {
+        aes_context =
+            std::unique_ptr<EVP_CIPHER_CTX, void (*)(EVP_CIPHER_CTX*)>(
+                crypto_shared->EVP_CIPHER_CTX_new(),
+                crypto_shared->EVP_CIPHER_CTX_free_ptr());
+      }
+
+      memcpy(iv.bytes, nonce_, AES_BLOCK_SIZE);
+      BigEndianAdd128(iv.bytes, block_index);
+
+      ret_val = crypto_shared->EVP_EncryptInit_ex(
+          aes_context.get(), crypto_shared->EVP_aes_256_ctr(), nullptr,
+          key_.key, iv.bytes);
+      if (1 == ret_val) {
+        out_len = 0;
+        ret_val = crypto_shared->EVP_EncryptUpdate(
+            aes_context.get(), (unsigned char*)data, &out_len,
+            (unsigned char*)data, (int)data_size);
+
+        if (1 != ret_val || (int)data_size != out_len) {
+          status = Status::InvalidArgument("EVP_EncryptUpdate failed: ",
+                                           (int)data_size == out_len
+                                               ? "bad return value"
+                                               : "output length short");
+        }
+        // this is a soft reset of aes_context per man pages
+        uint8_t temp_buf[AES_BLOCK_SIZE];
+        out_len = 0;
+        ret_val = crypto_shared->EVP_EncryptFinal_ex(aes_context.get(),
+                                                     temp_buf, &out_len);
+
+        if (1 != ret_val || 0 != out_len) {
+          status = Status::InvalidArgument("EVP_EncryptFinal_ex failed.");
+        }
+      } else {
+        status = Status::InvalidArgument("EVP_EncryptInit_ex failed.");
+      }
+    } else {
+      status = Status::NotSupported(
+          "libcrypto not available for encryption/decryption.");
+    }
+  }
+
+  return status;
+}
+
+// Decrypt one or more (partial) blocks of data at the file offset.
+//  Length of data is given in data_size.
+//  CTR Encrypt and Decrypt are synonyms.  Using Encrypt calls here to reduce
+//   count of symbols loaded from libcrypto.
+Status AESBlockAccessCipherStream::Decrypt(uint64_t file_offset, char* data,
+                                           size_t data_size) {
+  // Calculate block index
+  size_t block_size = BlockSize();
+  uint64_t block_index = file_offset / block_size;
+  size_t block_offset = file_offset % block_size;
+  size_t remaining = data_size;
+  size_t prefix_size = 0;
+  uint8_t temp_buf[block_size];
+
+  Status status;
+  ALIGN16 AesAlignedBlock iv;
+  int out_len = 0, ret_val;
+
+  if (crypto_shared->IsValid()) {
+    // make a context once per thread
+    if (!aes_context) {
+      aes_context = std::unique_ptr<EVP_CIPHER_CTX, void (*)(EVP_CIPHER_CTX*)>(
+          crypto_shared->EVP_CIPHER_CTX_new(),
+          crypto_shared->EVP_CIPHER_CTX_free_ptr());
+    }
+
+    memcpy(iv.bytes, nonce_, AES_BLOCK_SIZE);
+    BigEndianAdd128(iv.bytes, block_index);
+
+    ret_val = crypto_shared->EVP_EncryptInit_ex(
```

Comment:
This might be my lack of understanding of libcrypto, but shouldn't Decrypt need to call EVP_Decrypt?

Reply:
CTR mode is an exception to the rule.  Decrypt and Encrypt are the same.  Why?  Do not know.  Just following examples and seeing good data on the output side.